File: src/features/payments/data/payment.ts
export type PaymentStatus = 'processing' | 'success' | 'failed';

export interface Payment {
  readonly id: string;
  readonly billerId: string;
  readonly billerName: string;
  readonly categoryId: string;
  readonly account: string;
  readonly amountPaise: number;
  readonly paidAtUtc: Date;
  readonly status: PaymentStatus;
}

export interface Payments {
  readonly items: readonly Payment[];
  readonly nextId: number;
}

export const emptyPayments = (): Payments => ({ items: [], nextId: 1 });

export const byId = (payments: Payments, id: string): Payment | null =>
  payments.items.find((p) => p.id === id) ?? null;

export const hasProcessing = (payments: Payments, billerId: string, account: string): boolean =>
  payments.items.some(
    (p) => p.billerId === billerId && p.account === account && p.status === 'processing',
  );

export const adding = (payments: Payments, payment: Payment): Payments => ({
  items: [...payments.items, payment],
  nextId: payments.nextId + 1,
});

export const updatingStatus = (payments: Payments, id: string, status: PaymentStatus): Payments => ({
  ...payments,
  items: payments.items.map((p) => (p.id === id ? { ...p, status } : p)),
});

File: lib/features/payments/data/payment.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'payment.freezed.dart';

enum PaymentStatus { processing, success, failed }

@freezed
abstract class Payment with _$Payment {
  const factory Payment({
    required String id,
    required String billerId,
    required String billerName,
    required String categoryId,
    required String account,
    required int amountPaise,
    required DateTime paidAtUtc,
    required PaymentStatus status,
  }) = _Payment;
}

@freezed
abstract class Payments with _$Payments {
  const Payments._();

  const factory Payments({
    required List<Payment> items,
    required int nextId,
  }) = _Payments;

  factory Payments.empty() => const Payments(items: [], nextId: 1);

  Payment? byId(String id) => items.where((p) => p.id == id).firstOrNull;

  bool hasProcessing(String billerId, String account) => items.any((p) =>
      p.billerId == billerId &&
      p.account == account &&
      p.status == PaymentStatus.processing);

  Payments adding(Payment payment) =>
      Payments(items: [...items, payment], nextId: nextId + 1);

  Payments updatingStatus(String id, PaymentStatus status) => copyWith(
      items: [
        for (final p in items) p.id == id ? p.copyWith(status: status) : p
      ]);
}
